src/constraint.rs

```diff
@@ -977,3 +993,83 @@ impl ConstraintValidator for CodepointLengthConstraint {
         Ok(())
     }
 }
+
+/// Implements a `element` constraint
+/// [element]: https://amzn.github.io/ion-schema/docs/spec.html#element
+#[derive(Debug, Clone, PartialEq)]
+pub struct ElementConstraint {
+    type_id: TypeId,
+}
+
+impl ElementConstraint {
+    pub fn new(type_id: TypeId) -> Self {
+        Self { type_id }
+    }
+
+    /// Tries to create an element constraint from the given OwnedElement
+    pub fn resolve_from_isl_constraint(
+        type_reference: &IslTypeRef,
+        type_store: &mut TypeStore,
+        pending_types: &mut PendingTypes,
+    ) -> IonSchemaResult<Self> {
+        let type_id =
+            IslTypeRef::resolve_type_reference(type_reference, type_store, pending_types)?;
+        Ok(ElementConstraint::new(type_id))
+    }
+}
+
+impl ConstraintValidator for ElementConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        let mut violations: Vec<Violation> = vec![];
+
+        // Check for null container
+        if value.is_null() {
+            return Err(Violation::new(
+                "element",
+                ViolationCode::TypeMismatched,
+                &format!("expected a container found {:?}", value),
+            ));
+        }
+
+        let type_def = type_store.get_type_by_id(self.type_id).unwrap();
+
+        // validate each element of the given value container
+        match value.ion_type() {
+            IonType::List | IonType::SExpression => {
+                for val in value.as_sequence().unwrap().iter() {
+                    if let Err(violation) = type_def.validate(val, type_store) {
+                        violations.push(violation);
+                    }
+                }
+            }
+            IonType::Struct => {
+                for (field_name, val) in value.as_struct().unwrap().iter() {
+                    if let Err(violation) = type_def.validate(val, type_store) {
+                        violations.push(violation);
+                    }
+                }
+            }
```

Comment:
It seems like there's a lot of places where you have to have separate branches for `struct` and `sequence`. Should we open an issue in `ion-rust` to add a `container` concept so that we could simplify this to `match value.as_container()`?